src/dhtproto/client/request/internal/GetHashRange.d

```diff
@@ -208,229 +193,195 @@ public struct GetHashRange
     /***************************************************************************
 
-        Default running state. Receives and handles hash-range update messages
-        from the node.
-
-        Next state:
-            - Receiving again
-
-        Returns:
-            next state
+        Handler policy, called from AllNodesRequest template to run the
+        request's main handling logic.
 
     ***************************************************************************/
 
-    private State stateReceiving ( )
+    private void handle ( )
     {
-        bool msg_error;
+        // Receive and store node's current hash range
+        hash_t min, max;
         this.conn.receive(
-            ( in void[] const_payload )
+            ( in void[] payload )
             {
-                Const!(void)[] payload = const_payload;
-
-                auto msg_type =
-                    *this.conn.message_parser.getValue!(MessageType)(payload);
-
-                with ( MessageType ) switch ( msg_type )
-                {
-                    case NewNode:
-                        typeof(AddrPort.naddress) addr;
-                        typeof(AddrPort.nport) port;
-                        hash_t min, max;
-                        this.conn.message_parser.parseBody(payload,
-                            addr, port, min, max);
-
-                        auto shared_resources = SharedResources.fromObject(
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            AddrPort(addr, port), min, max);
-                        break;
-
-                    case ChangeHashRange:
-                        hash_t new_min, new_max;
-                        this.conn.message_parser.parseBody(payload,
-                            new_min, new_max);
-
-                        auto shared_resources = SharedResources.fromObject(
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            this.conn.remote_address, new_min, new_max);
-                        break;
-
-                    default:
-                        log.error("Unknown message code {} received", msg_type);
-                        this.conn.shutdownWithProtocolError("Message parsing error");
-                }
+                Const!(void)[] payload_slice = payload;
+                min = *this.conn.message_parser.getValue!(hash_t)(payload_slice);
+                max = *this.conn.message_parser.getValue!(hash_t)(payload_slice);
             }
         );
+        this.resources.node_hash_ranges.updateNodeHashRange(
+            this.conn.remote_address, min, max);
 
-        return State.Receiving;
+        while ( true )
+        {
+            this.conn.receive(
+                ( in void[] payload )
+                {
+                    Const!(void)[] payload_slice = payload;
```

Comment:
... why I asked the previous question is because this is (as expected) very common thing to do - to use `MessageParser` inside receive delegate.